### PR TITLE
Hedging strategy also deep-copies context for primary execution

### PR DESCRIPTION
**Pull request by martintmk**
# Pull Request

## The issue or feature being addressed

#1727

## Details on the issue fix or feature implementation

As per #1727 there was inconsistency on how the hedging contexts for primary executions are handled. Instead of creating deep-clone, the hedging strategy re-used the original context. 

With this change, the handling is the same for both primary and hedged actions. Hedging strategy always creates a deep clone in all scenarios.

This also simplifies the implementation and handling of contexts. 
## Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Review by martincostello — COMMENTED** [review]
Is there any impact to benchmarks (+ve or -ve) from this change?

**Comment by martintmk** [comment]
> Is there any impact to benchmarks (+ve or -ve) from this change?

I don't think so, but let me re-run just in case.

File: bench/BenchmarkDotNet.Artifacts/results/Polly.Core.Benchmarks.HedgingBenchmark-report-github.md [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-|           Hedging_Secondary | 2.253 μs | 0.0051 μs | 0.0075 μs |  1.57 |    0.01 | 0.0038 |     184 B |        4.60 |
-|   Hedging_Primary_AsyncWork | 3.903 μs | 0.0260 μs | 0.0381 μs |  2.73 |    0.03 | 0.0610 |    1636 B |       40.90 |
-| Hedging_Secondary_AsyncWork | 4.936 μs | 0.0424 μs | 0.0595 μs |  3.45 |    0.05 | 0.0687 |    1838 B |       45.95 |
+| Method                      | Mean      | Error     | StdDev    | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
+|---------------------------- |----------:|----------:|----------:|------:|--------:|-------:|----------:|------------:|
+| Hedging_Primary             |  1.284 μs | 0.0047 μs | 0.0066 μs |  1.00 |    0.00 |      - |      40 B |        1.00 |
+| Hedging_Secondary           |  2.007 μs | 0.0043 μs | 0.0064 μs |  1.56 |    0.01 | 0.0038 |     184 B |        4.60 |
+| Hedging_Primary_AsyncWork   | 33.379 μs | 1.9491 μs | 2.9173 μs | 26.43 |    1.90 | 0.6104 |   15299 B |      382.48 |
+| Hedging_Secondary_AsyncWork | 33.735 μs | 0.2915 μs | 0.4273 μs | 26.28 |    0.43 | 0.6104 |   15431 B |      385.77 |
```
Comment (martincostello) [inline bench/BenchmarkDotNet.Artifacts/results/Polly.Core.Benchmarks.HedgingBenchmark-report-github.md:17]: Is this noise, or did it really 26x the allocations?

Comment (martintmk) [reply bench/BenchmarkDotNet.Artifacts/results/Polly.Core.Benchmarks.HedgingBenchmark-report-github.md:17]: It on virtualized VM, sometimes it goes crazy when asyn is involved. The first 2 lines are important.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
@martintmk Would you consider this change to be a patch or a minor change?

If we take the change for AoT too, then can that be 8.0.1 or does this change make it 8.1.0?

I'm thinking 8.0.1, but if you consider this minor rather than patch then we can do an 8.1.0 and the .NET 8 release will become 8.2.0.

**Comment by martintmk** [comment]
We had some API additions in #1687 so I would say we should do 8.1.0.

**Pull request merged**